core/src/block.rs

```diff
@@ -421,109 +435,81 @@ mod signed {
-
-    /// Revalidate the block that was sent through the network by transforming it back into [`ValidBlock`]
-    #[version_with_scale(n = 1, versioned = "VersionedCandidateBlock")]
+    /// Candidate block that has to be revalidated
     #[derive(Debug, Clone, Decode, Encode)]
-    pub struct CandidateBlock {
-        /// Block header
-        pub header: BlockHeader,
-        /// Array of rejected transactions.
-        pub rejected_transactions: Vec<VersionedSignedTransaction>,
-        /// Array of all transactions in this block.
-        pub transactions: Vec<VersionedSignedTransaction>,
-        /// Signatures of peers which approved this block.
-        pub signatures: SignaturesOf<Self>,
-        /// Event recommendations.
-        pub event_recommendations: Vec<Event>,
-    }
-
-    impl VersionedCandidateBlock {
-        #[allow(missing_docs)]
-        #[inline]
-        pub const fn as_v1(&self) -> &CandidateBlock {
-            match self {
-                Self::V1(v1) => v1,
-            }
-        }
-
-        #[allow(missing_docs)]
-        #[inline]
-        pub fn as_mut_v1(&mut self) -> &mut CandidateBlock {
-            match self {
-                Self::V1(v1) => v1,
-            }
-        }
+    #[repr(transparent)]
+    pub struct Candidate<B>(B);
 
-        #[allow(missing_docs)]
+    impl Candidate<VersionedSignedBlock> {
+        /// Calculate the hash of the candidate block.
         #[inline]
-        pub fn into_v1(self) -> CandidateBlock {
-            match self {
-                Self::V1(v1) => v1,
-            }
+        pub fn hash(&self) -> HashOf<Self> {
+            self.0.as_v1().hash().transmute()
         }
 
-        /// Return the header of a valid block
+        /// Returns the header of a candidate block
         #[inline]
         pub const fn header(&self) -> &BlockHeader {
-            &self.as_v1().header
-        }
-
-        /// Calculate the hash of the current block.
-        #[inline]
-        pub fn hash(&self) -> HashOf<Self> {
-            self.as_v1().hash().transmute()
+            &self.0.as_v1().header
         }
 
         /// Return signatures that are verified with the `hash` of this block, removing all other signatures.
         #[inline]
-        pub fn retain_verified_signatures(&mut self) -> impl Iterator<Item = &SignatureOf<Self>> {
-            self.as_mut_v1()
-                .retain_verified_signatures()
-                .map(SignatureOf::transmute_ref)
+        pub fn retain_verified_signatures(
+            &mut self,
+        ) -> impl Iterator<Item = &SignatureOf<SignedBlock>> {
+            self.0.as_mut_v1().retain_verified_signatures()
         }
 
         /// Revalidate a block against the current state of the world.
         ///
         /// # Errors
-        /// Forward errors from [`CandidateBlock::revalidate`]
-        #[inline]
+        /// - Block is empty
+        /// - Block has committed transactions
```

Comment:
What's wrong with that? I don't get it

Reply:
Yes, exactly. We want to avoid double-spending and stuff like that.